pyanaconda/users.py

```diff
@@ -286,179 +251,166 @@ def createUser(self, user_name, *args, **kwargs):
-        return self.setUserPassword("root", password, isCrypted, isLocked, algo)
+        # Reset sp_lstchg to an empty string. On systems with no rtc, this
+        # field can be set to 0, which has a special meaning that the password
+        # must be reset on the next login.
+        iutil.execWithRedirect("chage", ["-R", root, "-d", "", username])
+
+    def setRootPassword(self, password, isCrypted=False, isLocked=False, algo=None, root="/"):
+        return self.setUserPassword("root", password, isCrypted, isLocked, algo, root)
 
     def setUserSshKey(self, username, key, **kwargs):
-        childpid = self._prepareChroot(kwargs.get("root", iutil.getSysroot()))
-
-        if childpid == 0:
-            user = self.admin.lookupUserByName(username)
-            if not user:
-                log.error("setUserSshKey: user %s does not exist", username)
-                os._exit(1)
-
-            homedir = user.get(libuser.HOMEDIRECTORY)[0]
-            if not os.path.exists(homedir):
-                log.error("setUserSshKey: home directory for %s does not exist", username)
-                os._exit(1)
-
-            sshdir = os.path.join(homedir, ".ssh")
-            if not os.path.isdir(sshdir):
-                os.mkdir(sshdir, 0o700)
-                iutil.eintr_retry_call(os.chown, sshdir, user.get(libuser.UIDNUMBER)[0], user.get(libuser.GIDNUMBER)[0])
-
-            authfile = os.path.join(sshdir, "authorized_keys")
-            authfile_existed = os.path.exists(authfile)
-            with iutil.open_with_perm(authfile, "a", 0o600) as f:
-                f.write(key + "\n")
-
-            # Only change ownership if we created it
-            if not authfile_existed:
-                iutil.eintr_retry_call(os.chown, authfile, user.get(libuser.UIDNUMBER)[0], user.get(libuser.GIDNUMBER)[0])
-                iutil.execWithRedirect("restorecon", ["-r", sshdir])
-            os._exit(0)
-        else:
-            return self._finishChroot(childpid)
+        root = kwargs.get("root", iutil.getSysroot())
+
+        pwent = self._getpwnam(username, root)
+        if not pwent:
+            raise ValueError("setUserSshKey: user %s does not exist" % username)
+
+        homedir = root + pwent[5]
+        if not os.path.exists(homedir):
+            log.error("setUserSshKey: home directory for %s does not exist", username)
+            os._exit(1)
+
+        uid = pwent[2]
+        gid = pwent[3]
+
+        sshdir = os.path.join(homedir, ".ssh")
+        if not os.path.isdir(sshdir):
+            os.mkdir(sshdir, 0o700)
+            iutil.eintr_retry_call(os.chown, sshdir, int(uid), int(gid))
+
+        authfile = os.path.join(sshdir, "authorized_keys")
+        authfile_existed = os.path.exists(authfile)
+        with iutil.open_with_perm(authfile, "a", 0o600) as f:
+            f.write(key + "\n")
+
+        # Only change ownership if we created it
+        if not authfile_existed:
```

Comment:
Maybe I missed something but I think we want to change ownership only when the out file "is" created.